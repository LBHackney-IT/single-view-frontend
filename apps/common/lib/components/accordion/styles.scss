@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-accordion/accordion";
@import "~govuk-frontend/govuk/helpers/visually-hidden";

.lbh-accordion--hide-controls .govuk-accordion__controls {
  @include govuk-visually-hidden;
}
