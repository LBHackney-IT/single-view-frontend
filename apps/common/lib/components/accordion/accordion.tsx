import React, { ReactElement, forwardRef, useEffect, useRef } from "react";
import mergeRefs from "react-merge-refs";

import cn from "classnames";
import { Accordion as AccordionJs } from "lbh-frontend";

import { widthOverrides } from "../../utils";
import { Heading } from "../heading";

import type * as Polymorphic from "@radix-ui/react-polymorphic";
import "./styles.scss";

export interface AccordionItemProps {
  id: string;
  title: string;
}

export type AccordionItemComponent = Polymorphic.ForwardRefComponent<
  "div",
  AccordionItemProps
>;

export const AccordionItem: AccordionItemComponent = forwardRef(function AccordionItem(
  { as: AccordionItemComp = "div", children, className, id, title },
  ref,
) {
  return (
    <AccordionItemComp ref={ref} className={cn("govuk-accordion__section", className)}>
      <div className="govuk-accordion__section-header">
        <Heading as="h3" variant="h5" className="govuk-accordion__section-heading">
          <span
            className="govuk-accordion__section-button"
            id={`accordion-heading-${id}`}
          >
            {title}
          </span>
        </Heading>
      </div>
      <div
        id={`accordion-content-${id}`}
        className="govuk-accordion__section-content"
        aria-labelledby={`accordion-heading-${id}`}
      >
        {children}
      </div>
    </AccordionItemComp>
  );
});

type AccordionChild =
  | ReactElement<AccordionItemProps>
  | ReactElement<AccordionItemProps>[]
  | null;

export interface AccordionProps {
  id: string;
  children: AccordionChild | AccordionChild[];
  defaultIndex?: number;
  visuallyHideControls?: boolean;
  override?: number;
}

export type AccordionComponent = Polymorphic.ForwardRefComponent<"div", AccordionProps>;

export const Accordion: AccordionComponent = forwardRef(function Accordion(
  {
    as: AccordionComp = "div",
    className,
    defaultIndex,
    override,
    visuallyHideControls = false,
    ...props
  },
  ref,
) {
  const localRef = useRef<HTMLElement>(null);
  const defaultIndexRef = useRef<number | undefined>(defaultIndex);

  useEffect(() => {
    /* istanbul ignore else */
    if (localRef.current) {
      const acc = new AccordionJs(localRef.current);
      acc.init();
      /* istanbul ignore else */
      if (defaultIndexRef.current !== undefined) {
        const section = acc.$sections.item(defaultIndexRef.current);
        /* istanbul ignore else */
        if (section) {
          const button = section.querySelector<HTMLButtonElement>(
            `.${acc.sectionButtonClass}`,
          );
          /* istanbul ignore else */
          if (button) {
            const contentId = button.getAttribute("aria-controls");
            /* istanbul ignore else */
            if (contentId && !window.sessionStorage.getItem(contentId)) {
              acc.setExpanded(true, acc.$sections.item(defaultIndexRef.current));
            }
          }
        }
      }
    }
  }, []);

  return (
    <AccordionComp
      className={cn(
        "govuk-accordion",
        "lbh-accordion",
        { "lbh-accordion--hide-controls": visuallyHideControls },
        widthOverrides(override),
        className,
      )}
      data-attribute="value"
      ref={mergeRefs([localRef, ref])}
      {...props}
    />
  );
});
