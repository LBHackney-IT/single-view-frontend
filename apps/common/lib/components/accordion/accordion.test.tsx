import React from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";
import { screen } from "@testing-library/react";

import { Accordion, AccordionItem } from "./accordion";

test("it passes a11y", async () => {
  const { container } = render(
    <Accordion id="test">
      <AccordionItem id="test-1" title="Test">
        Hello
      </AccordionItem>
      <AccordionItem id="test-2" title="Next">
        Hello
      </AccordionItem>
    </Accordion>,
  );

  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders correctly", () => {
  render(
    <Accordion id="test">
      <AccordionItem id="test-1" title="Test">
        Hello
      </AccordionItem>
      <AccordionItem id="test-2" title="Next">
        Hello
      </AccordionItem>
    </Accordion>,
  );

  expect(screen.getAllByRole("button")).toHaveLength(3);
  expect(screen.getByText("Test")).toBeInTheDocument();
});

test("it sets the defaultIndex to be expanded", () => {
  render(
    <Accordion id="test" defaultIndex={1}>
      <AccordionItem id="test-1" title="Test">
        Hello
      </AccordionItem>
      <AccordionItem id="test-2" title="Next">
        Hello
      </AccordionItem>
    </Accordion>,
  );

  const button = screen.getAllByRole("button")[2];
  expect(button?.getAttribute("aria-expanded")).toBe("true");
});
