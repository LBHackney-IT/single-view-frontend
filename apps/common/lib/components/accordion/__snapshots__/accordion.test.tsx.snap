// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it passes a11y 1`] = `
<div>
  <div
    class="govuk-accordion lbh-accordion"
    data-attribute="value"
    id="test"
  >
    <div
      class="govuk-accordion__controls"
    >
      <button
        aria-expanded="false"
        class="govuk-accordion__open-all"
        type="button"
      >
        Open all
        <span
          class="govuk-visually-hidden"
        >
           sections
        </span>
      </button>
    </div>
    <div
      class="govuk-accordion__section"
    >
      <div
        class="govuk-accordion__section-header"
      >
        <h3
          class="lbh-heading-h5 govuk-accordion__section-heading"
        >
          <button
            aria-controls="test-content-1"
            aria-expanded="false"
            class="govuk-accordion__section-button"
            id="accordion-heading-test-1"
            type="button"
          >
            Test
            <span
              aria-hidden="true"
              class="govuk-accordion__icon"
            />
          </button>
        </h3>
      </div>
      <div
        aria-labelledby="accordion-heading-test-1"
        class="govuk-accordion__section-content"
        id="accordion-content-test-1"
      >
        Hello
      </div>
    </div>
    <div
      class="govuk-accordion__section"
    >
      <div
        class="govuk-accordion__section-header"
      >
        <h3
          class="lbh-heading-h5 govuk-accordion__section-heading"
        >
          <button
            aria-controls="test-content-2"
            aria-expanded="false"
            class="govuk-accordion__section-button"
            id="accordion-heading-test-2"
            type="button"
          >
            Next
            <span
              aria-hidden="true"
              class="govuk-accordion__icon"
            />
          </button>
        </h3>
      </div>
      <div
        aria-labelledby="accordion-heading-test-2"
        class="govuk-accordion__section-content"
        id="accordion-content-test-2"
      >
        Hello
      </div>
    </div>
  </div>
</div>
`;
